### PR TITLE
[bcryptjs] update definitions for v2.4.0

### PR DESCRIPTION
**Pull request by RafaelKr**
Please fill in this template.

- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [x] Run `tsc` without errors.
- [x] Run `npm run lint package-name` if a `tslint.json` is present.

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/dcodeIO/bcrypt.js#api
- [x] Increase the version number in the header if appropriate.

### CODE REVIEW

File: bcryptjs/index.d.ts [hunk 1]
```diff
@@ -63,13 +70,16 @@ export declare function compareSync(s: string, hash: string): boolean;
...
+ * @throws If callback is present but not a function
  */
-export declare function compare(s: string, hash: string, callback: (err: Error, success: boolean) => void, progressCallback?: (percent: number) => void): void;
+export declare function compare(s: string, hash: string, callback?: (err: Error, success: boolean) => void, progressCallback?: (percent: number) => void): Promise<boolean>;
 
 /**
  * Gets the number of rounds used to encrypt the specified hash.
  * @param  hash Hash to extract the used number of rounds from
  * @return Number of rounds used
+ * @throws If hash is not a string
```
Comment (andy-ms) [inline bcryptjs/index.d.ts]: This is what type annotations are for. Most functions can throw an exception if they get a parameter of the wrong type.

File: bcryptjs/index.d.ts [hunk 2]
```diff
@@ -16,21 +16,26 @@ export declare function setRandomFallback(random: (random: number) => number[]):
...
 
 /**
  * Asynchronously generates a salt.
- * @param callback Callback receiving the error, if any, and the resulting salt
+ * @param  callback Callback receiving the error, if any, and the resulting salt
+ * @return Promise with resulting salt, if callback has been omitted
+ * @throws If callback is present but not a function
  */
-export declare function genSalt(callback: (err: Error, salt: string) => void): void;
+export declare function genSalt(callback?: (err: Error, salt: string) => void): Promise<string>;
```
Comment (andy-ms) [inline bcryptjs/index.d.ts]: Use 2 overloads, one taking a callback, one returning a promise.

Comment (RafaelKr) [inline bcryptjs/index.d.ts]: Thank you very much! This is my first definition and wasn't really sure, how to solve that. Thanks!

File: bcryptjs/index.d.ts (hunk 1 above)
Comment (RafaelKr) [inline bcryptjs/index.d.ts]: So you mean, I should remove this line? Copied it from the API docs.

File: bcryptjs/index.d.ts [hunk 3]
```diff
@@ -16,11 +16,19 @@ export declare function setRandomFallback(random: (random: number) => number[]):
  * Synchronously generates a salt.
  * @param  rounds Number of rounds to use, defaults to 10 if omitted
  * @return Resulting salt
+ * @throws If a random fallback is required but not set
```
Comment (RafaelKr) [inline bcryptjs/index.d.ts]: @andy-ms What about this throw-decorator? I'm asking, because this one is not related to the type annotations.
https://github.com/dcodeIO/bcrypt.js#gensaltsyncrounds-seed_length

Comment (andy-ms) [inline bcryptjs/index.d.ts]: That's fine. :+1:

**Comment by Joshua-F** [comment]
Andy seemed to have pointed out things you got them fixed up so this looks fine to me as well.

👍
